File: InterfaceMethodsPython/luCholesky.py
import numpy as np
import sys
from math import *


def readMatrix(file,size):
    L = np.zeros((size, size))
    U = np.zeros((size, size))
    A = np.zeros((size, size))
    B = np.zeros(size)
    Z = np.zeros(size)
    X = np.zeros(size)
    for i in range (size):
        cont=0
        fileMio=file.readline().replace("\n","").split(" ")
        linea=fileMio[0:size]
        for j in linea:
            A[i][cont]=float(j) 
            cont+=1
        valorB=fileMio[len(fileMio)-1]
        B[i]=float(valorB)

    return L,U,A,B,Z,X


def cholesky (L,U,A,n):    
    for k in range(n):
        suma1=0
        for m in range (k):
            suma1+=L[k][m]*U[m][k]
        L[k][k]=sqrt(A[k][k]-suma1)
        U[k][k]=L[k][k]

        for i in range (k,n):
            suma2=0
            for p in range (0,k):
                suma2+=L[i][p]*U[p][k]
            L[i][k]=(A[i][k]-suma2)/float(U[k][k])

        for j in range (k+1,n):
            suma3=0
            for h in range (k):
                suma3+=L[k][h]*U[h][j]
            U[k][j]=(A[k][j]-suma3)/float(L[k][k]) 
    return L,U

def progressive(Z,B,L,n):
    for i in range (n):
        sum=0.0
        for p in range(i):
            sum += L[i][p]*Z[p]
        Z[i]=(B[i]-sum)/L[i][i]
    return Z
def regressiveC(X,Z,U,n):
    for i in range (n-1,-1,-1):
        sum=0.0
        for p in range(i+1,n):
            sum += U[i][p]*X[p]
        X[i]=(Z[i]-sum)/U[i][i]
    return X

    
def main():
    name=sys.argv[1]
    
    file=open(name)
    size=int(file.readline())
    L,U,A,B,Z,X=readMatrix(file,size)
    
    L,U=cholesky(L,U,A,size)
    Z=progressive(Z,B,L,size)
    X=regressiveC(X,Z,U,size)
    print ("Matrix A")
    print (np.array(A))
    print ("\n")
    print ("Matrix L")
    print np.array(L)
    print ("\n")
    print ("Matrix U")
    print np.array(U)
    print ("\n")
    print ("Vector z")
    print np.array(Z)
    print ("\n")

    print ("Vector X")
    print np.array(X)

main()

File: PrincipalMethods/luCholesky.cpp
#include <bits/stdc++.h>
#include <math.h>
#include <time.h>
using namespace std;

double **L, **U, *B, *Z, *X, **A;

//With this method we initialize two matrixes (L and U) whit a specific size
void fillMatrix( int size){
    for (int i = 0; i <size; ++i){
        for (int j =0 ; j<size; ++j){
            L[i][j] =0;
        }
    }
    for (int i = 0; i <size; ++i){
        for (int j =0 ; j<size; ++j){
            U[i][j]=0 ;
        }
    }
}

//Here we read the matrix and we get the value of U and L matrixes
void readMatrix( int size,string origin){
    B = new double[size];
    U = new double *[size];
    A= new double *[size];
    for (int i = 0; i < size; ++i){
        U[i] = new double [size];
    }

    for (int i = 0; i < size; ++i){
        A[i] = new double [size];
    }

    L = new double*[size];

    for (int i = 0; i < size; ++i){
        L[i] = new double[size];
    }

    fillMatrix(size);
    ifstream read;
    read.open(origin.c_str());
    read >> B[0];
    for (int i = 0; i < size; ++i){
        for (int j = 0; j < size; ++j){
            read >> A[i][j];
        }
        read >> B[i];
    }

    read.close();
}

//Here we try to solve the matrix with the method cholesky
void cholesky(int n){
    double suma1,suma2,suma3;

    for(int k=0;k<n;++k){
        suma1=0;
        for(int m=0;m<k;++m){
            suma1+=L[k][m]*U[m][k];
        }
        L[k][k]=sqrt(A[k][k]-suma1);
        U[k][k]=L[k][k];

        for(int i=k;i<n;++i){
            suma2=0;
            for(int p=0;p<k;++p){
                suma2+=L[i][p]*U[p][k];
            }
            L[i][k]=(A[i][k]-suma2)/(double)U[k][k];
        }

        for(int j=k+1;j<n;++j){
            suma3=0;
            for(int h=0;h<k;++h){
                suma3+=L[k][h]*U[h][j];
            }
            U[k][j]=(A[k][j]-suma3)/(double)L[k][k];

        }

    }
}

//Here we find the value of the Z's and then with this we find the value of the unknowns
void progressiveC(int n){
	Z = new double[n];
	for(int i=0; i<n; i++){
		double sum = 0;
		for(int p=0; p<i; ++p){
			sum += L[i][p]*Z[p];
		}
		Z[i]=(B[i]-sum)/L[i][i];
	}
}

//Here we find the value of the unknowns
void regressiveC(int n){
	X = new double[n];
	for(int i=n-1; i>=0; --i){
		double sum = 0;
		for(int p=i+1; p<n; ++p){
			sum += U[i][p]*X[p];
		}
		X[i]=(Z[i]-sum)/U[i][i];
	}
}

//Here we put the solution in a file
void printSolution(string solutionFile, int size){
    ofstream write;
    write.open(solutionFile.c_str(),ios::trunc);
    for (int i = 0; i <size; ++i){
        write << X[i] << endl;
    }


    write.close();
}

//This is the main method that ejecuted all the methods of the class
int main(){
	string originalFile,readingName,writingName,solutionFile;
    long long matrixSize;
    clock_t start, end;
    originalFile="matrix.txt";
    solutionFile="solution.txt";
    ifstream f("matrix.txt");
    f >> matrixSize;


    readMatrix(matrixSize,originalFile);
    start = clock();
    cholesky(matrixSize);
    progressiveC(matrixSize);
    regressiveC(matrixSize);
    end = clock();
    printf("The time was: %.30g\n", (double)( (end - start) / 1000.0));
    printSolution(solutionFile, matrixSize);
	return 0;
}
